lib/policies/throttlingRetryPolicy.ts

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+import { BaseRequestPolicy, RequestPolicy, RequestPolicyOptions, RequestPolicyFactory } from "./requestPolicy";
+import { WebResource } from "../webResource";
+import { HttpOperationResponse } from "../httpOperationResponse";
+import { Constants } from "../util/constants";
+
+type ResponseHandler = (httpRequest: WebResource, response: HttpOperationResponse) => Promise<HttpOperationResponse>;
+const StatusCodes = Constants.HttpConstants.StatusCodes;
+
+export function throttlingRetryPolicy(): RequestPolicyFactory {
+  return {
+    create: (nextPolicy: RequestPolicy, options: RequestPolicyOptions) => {
+      return new ThrottlingRetryPolicy(nextPolicy, options);
+    }
+  };
+}
+
+export class ThrottlingRetryPolicy extends BaseRequestPolicy {
+  private _handleResponse: ResponseHandler;
+
+  constructor(nextPolicy: RequestPolicy, options: RequestPolicyOptions, _handleResponse?: ResponseHandler) {
+    super(nextPolicy, options);
+    this._handleResponse = _handleResponse || this._defaultResponseHandler;
+  }
+
+  public async sendRequest(httpRequest: WebResource): Promise<HttpOperationResponse> {
+    return this._nextPolicy.sendRequest(httpRequest.clone()).then(response => {
+      if (response.status !== StatusCodes.TooManyRequests) {
+        return response;
+      } else {
+        return this._handleResponse(httpRequest, response);
+      }
+    });
+  }
+
+  private async _defaultResponseHandler(httpRequest: WebResource, httpResponse: HttpOperationResponse): Promise<HttpOperationResponse> {
+    const retryAfterHeader: string | undefined = httpResponse.headers.get(Constants.HeaderConstants.RETRY_AFTER);
+
+    if (retryAfterHeader) {
+      const delayInMs: number | undefined = ThrottlingRetryPolicy.parseRetryAfterHeader(retryAfterHeader);
+      if (delayInMs) {
+        return delay(delayInMs).then(_ => this._nextPolicy.sendRequest(httpRequest));
+      }
+    }
+
+    return httpResponse;
+  }
+
+  public static parseRetryAfterHeader(headerValue: string): number | undefined {
```

Comment:
Instead of making this a static function, I'd recommend just moving it outside of the class altogether.